### PR TITLE
#170947663 User will be able to view and edit profile information

### PR DESCRIPTION
**Pull request by EugeneMunya**
#### What does this PR do?
- Adding functionality of allowing user to view his/her profile information and edit them.
#### Description of Task to be completed?
- A user will have a profile page that will display  information like `names`, `email`, `gender`, `passport number`, `department`, `preferred currency`, `preferred language`, `user role`, `where you live`, `birthday`, `line manager`,` biography` and upload `profile image`. He/she will be able to view his/her own information and be able to edit them.
#### How should this be manually tested?
- open your terminal and run this command `git clone https://github.com/andela/knightsbn    backend.git` to clone repository
- run `cd knights-bn-backend` to get inside the project
- run `git fetch && git checkout ft-profile-page-settings-170947663` 
- run `npm install` to install dependencies
- check if you have PostgresSQL database if you don't have it install it from here [PostgreSQL](https://www.2ndquadrant.com/en/blog/pginstaller-install-postgresql/)
- create database
- create `.env` file in your project root directory
- follow `env.example` file to create environment variables in `.env` file and 
   assign them with all your database configurations
- run `npm run createTables` to create tables in your new database you have created.
- run `npm run test` to run all the tests
- run `npm run start` to start the server.
- use swagger documentation to test every end point by visiting this link `http://localhost:4000/api/v1/docs` on your browser or follow the next steps
- open your postman or any other API tester you have
-  to view or edit user profile you must be logged in or signed up , see how to do [log in](https://github.com/andela/knights-bn-backend/pull/23) or [sign up](https://github.com/andela/knights-bn-backend/pull/17)
- to test view profile information send GET request to this link `http://localhost:4000/api/v1/user/profile`
- to test update profile information send PATCH request to this link `http://localhost:4000/api/v1/edit/user/profile` and use form-data option under the body in your postman in order to find a way you can upload a profile image  and click on `Bulk Edit` option on the right part under save button and past the information bellow to generate a well define form,  after pasting the data click on `Key-Value Edit` to generate a form,  and fill the fields and remember to change profileImage field type to file so that you can upload image. after filling every field send your request.
```
firstName:eugene
lastName:rutikanga
email:eukigali@gmail.com
passport:12348934
gender:male
birthDay:1990-2-3
language:English
currency:Euro
homeTown:dallas
department:IT
lineManager:Peter Doe
biography:Lorem Ipsum is simply dummy text of the printing and typesetting industry
```

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#170947663](https://www.pivotaltracker.com/story/show/170947663)

#### Screenshots
get profile information

Screen Shot 2020-02-24 at 22 30 46

update the information

Screen Shot 2020-02-24 at 22 24 25

### CODE REVIEW

**Comment by Niyonsengaeric** [comment]
You forget the step of `logging in' into the system

**Comment by Niyonsengaeric** [comment]
Screen Shot 2020-02-20 at 15 23 08

You forget also to validate the `birthday` as you can see in the picture, it is saved while the year is 2024

**Comment by EugeneMunya** [comment]
> You forget the step of `logging in' into the system

I put it down let me put it  before the send step

**Comment by Niyonsengaeric** [comment]
it will be better if you arrange well the steps of testing your PR.

ex: as the title says you view and edit the profile, so the viewing should be on top before editing, in your PR description.

**Comment by Kagorora** [comment]
@EugeneMunya  In your PR description, It would be best If you included the PATCH in this line as follow

for updating user use  action `PATCH and use the following URL: `http://localhost:4000/api/v1/edit/user/profile`

File: src/db/migrations/20200208090739-create-user.js [hunk 1]
```diff
@@ -33,14 +33,50 @@ module.exports = {
...
       allowNull: false,
       type: Sequelize.DATE,
     },
+    createdAt: {
+      allowNull: false,
+      type: Sequelize.DATE,
+    }
   }),
-  down: (queryInterface, Sequelize) => queryInterface.dropTable('Users'),
+  down: (queryInterface, Sequelize) => queryInterface.dropTable('user'),
```
[the remark this answers is not in the document]
Comment (Kagorora) [reply src/db/migrations/20200208090739-create-user.js]: @EugeneMunya in future do not  alter an existing migration rather create another migration that will add your desired changes to an existing one.

File: src/tests/mainTest.test.js [hunk 2]
```diff
@@ -1,5 +1,7 @@
 import signUpTest from './user.test';
 import userLoginTest from './userLogin.test';
+import userProfileInformation from './getUserAnEditProfileInfo.test';
 
 signUpTest();
 userLoginTest();
+userProfileInformation();
```
Comment (Kagorora) [inline src/tests/mainTest.test.js:9]: @EugeneMunya  why haven't you tested this endpoint as well?

`router.get('/user/profile', auth.auth, userProfile.getProfileInformation);`

Screen Shot 2020-02-22 at 09 39 42

**Review by Kagorora — CHANGES_REQUESTED** [review]
@EugeneMunya  Please work on the feedback that I left and next time try to only change files that need to be changed.

**Comment by EugeneMunya** [comment]
> @EugeneMunya Please work on the feedback that I left and next time try to only change files that need to be changed.

**Comment by EugeneMunya** [comment]
d

File: src/tests/mainTest.test.js (hunk 2 above)
Comment (EugeneMunya) [reply src/tests/mainTest.test.js:9]: I'm still working on it

Comment (EugeneMunya) [reply src/tests/mainTest.test.js:9]: I managed to test everything, now you can check

**Comment by EugeneMunya** [comment]
> it will be better if you arrange well the steps of testing your PR.
> 
> ex: as the title says you view and edit the profile, so the viewing should be on top before editing, in your PR description.

I managed to organise everything well check now

**Comment by EugeneMunya** [comment]
> @EugeneMunya Please work on the feedback that I left and next time try to only change files that need to be changed.

I managed to do it you can see now

**Review by Kagorora — APPROVED** [review]

**Comment by Moise1** [comment]
@EugeneMunya ,  good stuff.

**Review by Moise1 — APPROVED** [review]

**Comment by T2Wil** [comment]
Hello @EugeneMunya ,
You did a nice job but I think you missed something while providing a profileImage link in the body response.
image


Try adding a `localhost:/4000/` in front of the profile image to make it clickable.

**Comment by EugeneMunya** [comment]
> Screen Shot 2020-02-20 at 15 23 08
> 
> You forget also to validate the `birthday` as you can see in the picture, it is saved while the year is 2024

I have validated the birthDay date you can check

**Comment by EugeneMunya** [comment]
> Hello @EugeneMunya ,
> You did a nice job but I think you missed something while providing a profileImage link in the body response.
> image
> 
> Try adding a `localhost:/4000/` in front of the profile image to make it clickable.

I have corrected it now it sade as a link you can check

**Comment by EugeneMunya** [comment]
> You forget the step of `logging in' into the system

It is there just keep reading you will see it

**Review by T2Wil — APPROVED** [review]

File: src/db/config/config.js [hunk 3]
```diff
@@ -9,13 +9,15 @@ module.exports = {
     database: process.env.DB_NAME,
     host: process.env.DB_HOST,
     dialect: 'postgres',
+    logging: false
```
Comment (MuhanguziDavid) [inline src/db/config/config.js]: Logging should be allowed for `development`

File: src/db/config/config.js [hunk 4]
```diff
@@ -24,5 +26,6 @@ module.exports = {
     host: process.env.HOST,
     url: process.env.DATABASE_URL,
     dialect: 'postgres',
+    logging: false
```
Comment (MuhanguziDavid) [inline src/db/config/config.js:29]: Logging should be allowed for `production`

File: src/db/migrations/20200213181730-create-user.js [hunk 5]
```diff
@@ -1,48 +1,49 @@
...
-      updatedAt: {
-        allowNull: false,
-        type: Sequelize.DATE
-      }
-    });
-  },
-  down: (queryInterface, Sequelize) => {
-    return queryInterface.dropTable('Users');
-  }
+  up: (queryInterface, Sequelize) => queryInterface.createTable('users', {
```
Comment (MuhanguziDavid) [inline src/db/migrations/20200213181730-create-user.js]: I noticed that you have two `users/Users` tables. Please rectify this.

Comment (EugeneMunya) [reply src/db/migrations/20200213181730-create-user.js]: I have managed to correct this mistake you can check

File: src/db/config/config.js (hunk 4 above)
Comment (EugeneMunya) [reply src/db/config/config.js:29]: `logging`option that disables Sequelize logging has been taken out in development mode

**Review by MuhanguziDavid — APPROVED** [review]
LGTM

**Pull request merged**